### PR TITLE
specs: add da spec

### PR DESCRIPTION
**Pull request by tuxcanfly**
## Overview

This package adds specifications for the DA interface and celestia implementation.

It also includes the failure modes when submitting rollup blocks to the da layer.

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

## Summary by CodeRabbit

- **New Features**
	- Introduced a Data Availability Layer Client (DALC) interface for improved interaction with Data Availability layers.
	- Implemented methods for submitting and retrieving blocks within the DALC.
	- Added gRPC service communication with protobuf serialization for rollup blocks.

- **Documentation**
	- Provided detailed descriptions and assumptions for interacting with the Data Availability layer.
	- Outlined failure cases and retry mechanisms for robust error handling.

### CODE REVIEW

**Comment by gupadhyaya** [comment]
@tuxcanfly rollkit's da will be the consumer of go-da interface and it will still have dalc as top level interface. can you adjust the specs from the consumer perspective? the current spec looks more like the go-da spec.

File: da/da.md [hunk 1]
```diff
@@ -1 +1,81 @@
...
+
+## Assumptions and Considerations
+
+When trying to submit a block to the Celestia, the following failure cases are possible:
+
+* submitting the Celestia transaction (rollup Block) into the mempool, but is not accepted for some reason e.g. mempool is full
+* submitting the Celestia transaction (rollup Block) into the mempool, but there is a nonce clash with a different transaction but the same nonce but different transaction. Example PFB submission plus TIA transfer using the same nonce. The one from rollkit will fail and never be included
+* submitting the Celestia transaction (rollup Block) but the gas was misconfigured too low so it will be stuck in the mempool
+* we call submit but get no response (node crashed for example)
+* we submit a tx but it is impossible to fit as it is too big (too many bytes) so it fails
```
Comment (Manav-Aggarwal) [inline da/da.md:63]: All references to Celestia should be changed to DA Layer since Rollkit aims to be DA Layer agnostic.

**Review by Manav-Aggarwal — CHANGES_REQUESTED** [review]

File: da/da.md (hunk 1 above)
Comment (tuxcanfly) [reply da/da.md]: Done.

File: da/da.md [hunk 2]
```diff
@@ -1 +1,81 @@
...
+The gRPC service uses protobufs for communication between the rollup and the Data Availability layer.
+
+Rollup blocks are serialised into binary as protobufs as per the encoding/decoding scheme defined in [block serialization].
+
+## Assumptions and Considerations
+
+When trying to submit a block to the Data Availability layer, the following failure cases are possible:
+
+* submitting the Data Availability layer transaction (rollup Block) into the mempool, but is not accepted for some reason e.g. mempool is full
+* submitting the Data Availability layer transaction (rollup Block) into the mempool, but there is a nonce clash with a different transaction but the same nonce but different transaction. Example PFB submission plus TIA transfer using the same nonce. The one from rollkit will fail and never be included
```
Comment (Manav-Aggarwal) [inline da/da.md:60]: Not a coherent sentence. Also, can you remove the reference to TIA here?

**Review by Manav-Aggarwal — CHANGES_REQUESTED** [review]

File: da/da.md (hunk 2 above)
Comment (tuxcanfly) [reply da/da.md]: Fixed.

**Review by Manav-Aggarwal — DISMISSED** [review]

File: da/da.md [hunk 3]
```diff
@@ -1 +1,81 @@
...
+
+When trying to submit a block to the Data Availability layer, the following failure cases are possible:
+
+* submitting the Data Availability layer transaction (rollup Block) into the mempool, but is not accepted for some reason e.g. mempool is full.
+* submitting the Data Availability layer transaction (rollup Block) into the mempool, but there is a nonce clash with a different transaction such as a coin transfer. The one from rollkit might fail and never be included.
+* submitting the Data Availability layer transaction (rollup Block) but the gas was misconfigured too low so it will be stuck in the mempool.
+* we call submit but get no response (node crashed for example).
+* we submit a tx but it is impossible to fit as it is too big (too many bytes) so it fails.
+
+In case of an error, the request will be retried with an exponentially increasing back-off starting with 100ms and capped at the DA block time, a maximum of 30 times until the request is successful.
```
Comment (nashqueue) [inline da/da.md:65]: This issue https://github.com/rollkit/rollkit/issues/1236 was originally a question. 

All those points are ideas that i had that still need to be validated of what actually causes a failure and what not. Relevant  discussion in here https://github.com/rollkit/rollkit/pull/1202#discussion_r1338665763

**Review by nashqueue — COMMENTED** [review]
Thank you . We need to be more precise of what failure and success means

**Review by gupadhyaya — APPROVED** [review]

**Comment by gupadhyaya** [comment]
Fixes #1236

**Pull request merged**